### ISSUE TITLE
ProMicro RP2040 dont work

### ISSUE THREAD

**Issue by pioz** [mohoyt/lagom#7]
Hi, I've a Lagom with a ATmega and works perfectly.
Now, I've switch with a new RP2040 (photo to see my setup).

![photo_2024-12-18_18-40-10](https://github.com/user-attachments/assets/b1c90c1c-4450-469d-a171-334def34a163)
![photo_2024-12-18_18-40-13](https://github.com/user-attachments/assets/ec112232-f13f-44fc-95f9-b15056943e8c)
![photo_2024-12-18_18-40-14](https://github.com/user-attachments/assets/69bf70ad-e912-421c-8c05-74ea44ded22e)
![photo_2024-12-18_18-40-15](https://github.com/user-attachments/assets/e1489709-1f48-4235-91c9-d5915f262152)

In the new micro RP2040 I've installed this firmware: https://github.com/mohoyt/lagom/releases/download/v1.0.2/sthlmkb_lagom_vial-bongocat_helios.uf2

Unfortunately, many of the keys are not functioning, and some keys print two characters instead of one. However, the mini OLED display works fine. Essentially, the keyboard does not work with the new microcontroller.

Did I miss something? Why am I encountering this issue? I also tested it with another Helios, and I encountered the same problem. Please help!

**Comment by mohoyt** [comment]
Hey Enrico,

That's super strange! I think I tested it with the same controller so hopefully it's not the firmware. 

Is there a pattern to the keys that don't work or are doubled up when you press? That might give us a clue on where to look for issues.

**Comment by pioz** [comment]
Sure, here my report:
```
8 -> 8 0
9 -> 9 -
i -> i p
o -> o [
] -> ]
\ -> \
k -> k ;
l -> l '
m -> m .
, -> , /
RETURN -> RETURN MUTE
UP -> UP
DOWN -> DOWN
LEFT -> LEFT RIGHT
```
Right rotary change volume, no click work

All other keys does nothing.

**Comment by pioz** [comment]
I've also tried with https://github.com/0xCB-dev/0xCB-Helios and I have same output.

**Comment by mohoyt** [comment]
Interesting! Sounds like there's some sort of short or interaction between cols 8 and 10 and 9 and 11 on the top IC. But strange that this doesn't happen on the Atmel pro micro. That still works if you plug it back in again right?

If you have a multi-meter can you see that the both ICs are getting power when the board is powered (pin 16 (top left of each IC) should be 3.3 to 5v) and that there are no shorts between the IC pins on the top row?

And does it work fine in Vial when you open it to configure the keymap?

**Comment by pioz** [comment]
If I put back the Atmel all works perfectly. I don't have in home a multi-meter. But I can borrow it to try.

**Comment by pioz** [comment]
@mohoyt Since I am not an expert in electronics, could you kindly tell me the voltage I should expect on each pin in the image by placing the positive probe of the multimeter on the pin and the negative probe on a GND pin?  Specifically, I am also interested in the pins of the SN74HC138N demultiplexer ICs.

![lagom](https://github.com/user-attachments/assets/d0146c35-8f50-4be0-a61f-ab7015cfce3c)

**Comment by pioz** [comment]
Ok thanks very much! 🙏🏻 How can I check the continuity from the pins to check if they are connected together?

**Comment by mohoyt** [comment]
Depending on the multimeter that you're testing with it might have a continuity mode (that would beep if there's a connection) but you can always set it to a resistance mode like 2000 ohms and then measure to see if there is a connection. 

If there's a connection and therefore continuity it will read 0, otherwise it will likely be max or 1 if there's not continuity. And you want to do this with the board disconnected from the computer.

**Comment by mohoyt** [comment]
Do you have an LED strip installed? Is there a way that you could temporarily disconnect it to see if that might be the issue? It could be drawing too much current from the microcontroller causing it to crash/behave flakily.

**Comment by pioz** [comment]
Yes, but I've already tried to remove the strip led. The problem persist.

**Comment by mohoyt** [comment]
I'm really struggling to think what it might be then. I will build up a board with this microcontroller later today to see if I get the same behaviour, but I'm fairly sure that it worked fine before when I created this firmware. 

Long shot, but are you sure that the rp2040 controllers are working 100%? Do you have another board or another way of testing that all the IO is working on them?

**Comment by pioz** [comment]
> I'm really struggling to think what it might be then. I will build up a board with this microcontroller later today to see if I get the same behaviour, but I'm fairly sure that it worked fine before when I created this firmware. 
> 
> 
> 
> Long shot, but are you sure that the rp2040 controllers are working 100%? Do you have another board or another way of testing that all the IO is working on them? 

Sure! Here’s an improved version:

I’ve been using the ATmega for months without any issues. However, the RP2040 isn’t working as expected as described before. I’ve ordered another board to start building a new keyboard from scratch. I’ll keep you updated.

**Comment by MisanZx** [comment]
Hi there all, I have same issue of doubling keys on the right hand side of the keyboard. left hand side does not work. oled is working fine.

**Comment by mohoyt** [comment]
Hey @MisanZx - Have you had a look at the debugging guide here: https://sthlmkb.com/2023/06/debugging-your-keyboard/ 

There might be an issue with the ICs, have you tried swapping them to see if that changes the behaviour?

**Comment by pioz** [comment]
The new PCB has just arrived. I tested continuity on the bare board with a multimeter, and everything checks out except for pin F4 (encC), which seems to be disconnected. Is this expected? According to the schematic, pin F4 should be connected to the encC hole (the top-right hole on the PCB).

I've attached a video, please make sure to play it with audio enabled:

https://github.com/user-attachments/assets/e92f5456-bb6f-4df8-9944-d8ab7c2e9947

**Comment by pioz** [comment]
Looking closely at the underside, there's a scratch, could that be the cause?


![photo_2025-01-03_21-09-35](https://github.com/user-attachments/assets/4a3e909c-3a3d-42f6-a915-e5efc2f2e33c)

**Comment by MisanZx** [comment]
I dont have any spare IC's I did swap them over to see if the other half works, no change.

**Comment by mohoyt** [comment]
@pioz That scratch could well be the cause. Pin F4 is only connected to that specific encoder pin and nothing else. So if it's not working then it's likely something on that trace. 

I wouldn't continue building that PCB then as you won't have a working encoder there unless you add a bodge wire (though a switch at that location would still work fine). Would you like me to send out a replacement right away for you?

**Comment by mohoyt** [comment]
@MisanZx Would you be able to reach out to hej@sthlmkb.com and we can send you some replacement ICs to swap in to hopefully solve your problem (or at least diagnose a little better what is happening) 😅

**Comment by pioz** [comment]
> @pioz That scratch could well be the cause. Pin F4 is only connected to that specific encoder pin and nothing else. So if it's not working then it's likely something on that trace. 
> 
> 
> 
> I wouldn't continue building that PCB then as you won't have a working encoder there unless you add a bodge wire (though a switch at that location would still work fine). Would you like me to send out a replacement right away for you?

Yes please 🙏🏻

**Comment by MisanZx** [comment]
@mohoyt no worries, thanks for the offer, I accidentally broke my own pcb board somewhere else now. Assuming it was the cheap microcontroller that was not very good as it was exactly the same one @pioz has. Thanks for your help anyways. I am building the board again with another friend soon but with a different micro controller so I should not run into any issues.

**Comment by mohoyt** [comment]
@pioz will send that first thing Tuesday (as Monday is a holiday here) - hope it won't be too long a wait! 

@MisanZx hopefully the other build goes better, but let me know if there are any issues and maybe I can help!

**Comment by pioz** [comment]
I've just finished building the new Lagom keyboard. I can finally say that everything works very well with the Pro Micro ATMega, RP2040, and also with [0xCB-Helios](https://github.com/0xCB-dev/0xCB-Helios). 💪🏻🍺"

Thanks @mohoyt for your help and support.
